src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -517,135 +168,287 @@ namespace cublog
+		//  - already passed on to the reusable job container
+		//  - dangling, as they have deleted themselves
+		jobs_vec->clear ();
+		m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_EXECUTE_AND_RETIRE);
 	      }
 	  }
       }
 
+    assert (jobs_vec->empty ());
+
     m_task_state_bookkeeping.set_inactive ();
+    m_in_execution.store (false);
   }
 
-  void redo_parallel::redo_task::retire ()
+  void
+  redo_parallel::redo_task::retire ()
   {
     // avoid self destruct, will be deleted by owning class
   }
 
-  void redo_parallel::redo_task::log_perf_stats () const
+  void
+  redo_parallel::redo_task::log_perf_stats () const
   {
     std::stringstream ss;
     ss << "Log recovery redo worker thread " << m_task_idx << " perf stats";
     m_perf_stats.log (ss.str ().c_str ());
   }
 
-  void redo_parallel::redo_task::accumulate_perf_stats (
+  void
+  redo_parallel::redo_task::accumulate_perf_stats (
 	  cubperf::stat_value *a_output_stats, std::size_t a_output_stats_size) const
   {
     m_perf_stats.accumulate (a_output_stats, a_output_stats_size);
   }
 
+  bool
+  redo_parallel::redo_task::is_idle () const
+  {
+    bool is_vector_empty = false;
+    {
+      std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
+      is_vector_empty = m_produce_vec->empty ();
+    }
+    return is_vector_empty && !m_in_execution.load ();
+  }
+
+  inline void
+  redo_parallel::redo_task::push_job (redo_parallel::redo_job_base *a_job)
+  {
+    std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
+
+    if (m_do_monitor_not_applied_log_lsa && m_produce_vec->empty ())
+      {
+	const log_lsa &job_log_lsa = a_job->get_log_lsa ();
+	set_not_applied_log_lsa_from_push_side (job_log_lsa);
+      }
+
+    m_produce_vec->push_back (a_job);
+  }
+
+  inline void
+  redo_parallel::redo_task::pop_jobs (std::unique_ptr<redo_parallel::redo_task::redo_job_vector_t> &a_out_job_vec)
+  {
+    assert (a_out_job_vec->empty ());
+
+    std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
+    m_produce_vec.swap (a_out_job_vec);
+
+    if (m_do_monitor_not_applied_log_lsa)
+      {
+	const log_lsa new_log_lsa =  a_out_job_vec->empty ()
+				     ? MAX_LSA
+				     : (*a_out_job_vec->begin ())->get_log_lsa ();
+	set_not_applied_log_lsa_from_pop_side (new_log_lsa);
+      }
+  }
+
+  inline void
+  redo_parallel::redo_task::set_not_applied_log_lsa_from_push_side (const log_lsa &a_log_lsa)
+  {
+    assert (m_do_monitor_not_applied_log_lsa);
+
+    {
+      const log_lsa snapshot_not_applied_log_lsa { m_not_applied_log_lsa.load () };
+      // strict comparison because jobs have ever-increasing log_lsa's
+      assert (snapshot_not_applied_log_lsa == MAX_LSA || snapshot_not_applied_log_lsa < a_log_lsa);
+    }
+
+    log_lsa expected_not_applied_log_lsa { MAX_LSA };
+    m_not_applied_log_lsa.compare_exchange_strong (expected_not_applied_log_lsa, a_log_lsa);
+    // TODO: any check for the return value?
```

Comment:
`m_not_applied_log_lsa = MAX_LSA` is set only with the protection of m_produce_vec_mtx and once set, it may not be changed back. The `compare_exchange_strong` may not be necessary.
```suggestion
    if (m_not_applied_log_lsa == MAX_LSA)
      {
        m_not_applied_log_lsa = MAX_LSA;
      }
```

Reply:
Removed.